runtime/compiler/optimizer/MethodHandleTransformer.cpp

```diff
@@ -773,3 +776,155 @@ TR_MethodHandleTransformer::process_java_lang_invoke_Invokers_checkCustomized(TR
+
+         uint32_t mhTableOffset = comp()->fej9()->getInstanceFieldOffsetIncludingHeader("Ljava/lang/invoke/VarHandle$TypesAndInvokers;", "methodHandle_table", "[Ljava/lang/invoke/MethodHandle;", comp()->getCurrentMethod());
+         const char * mhTableFieldNameAndSig = "java/lang/invoke/VarHandle$TypesAndInvokers.methodHandle_table [Ljava/lang/invoke/MethodHandle;";
+         TR::Node *mhTableHolder = tisNode;
+#else
+         uint32_t mhTableOffset = comp()->fej9()->getInstanceFieldOffsetIncludingHeader("Ljava/lang/invoke/VarHandle;", "methodHandleTable", "[Ljava/lang/invoke/MethodHandle;", comp()->getCurrentMethod());
+         const char * mhTableFieldNameAndSig = "java/lang/invoke/VarHandle.methodHandleTable [Ljava/lang/invoke/MethodHandle;";
+         TR::Node *mhTableHolder = node->getFirstArgument();
+#endif /* JAVA_SPEC_VERSION <= 17 */
+         TR::SymbolReference *mhTableSymRef = comp()->getSymRefTab()->findOrFabricateShadowSymbol(comp()->getMethodSymbol(),
+                                                                                 TR::Symbol::Java_lang_invoke_VarHandle_methodHandleTable,
+                                                                                 TR::Address,
+                                                                                 mhTableOffset,
+                                                                                 false,
+                                                                                 true,
+                                                                                 true,
+                                                                                 mhTableFieldNameAndSig);
+         TR::Node *mhTableNode = TR::Node::createWithSymRef(comp()->il.opCodeForIndirectLoad(TR::Address), 1, 1, mhTableHolder, mhTableSymRef);
+         mhTableNode->copyByteCodeInfo(node);
+
+         // get index for load
+         uintptr_t accessDescriptorObj = knot->getPointer(adIndex);
+         int32_t mhEntryIndex = comp()->fej9()->getInt32Field(accessDescriptorObj, "mode");
+         TR::Node *arrayIndexNode = TR::Node::create(TR::iconst, 0, mhEntryIndex);
+         arrayIndexNode->copyByteCodeInfo(node);
+
+         // calculate element address, create load from array, and refine MH symref with known object info
+         TR::Node *mhAddressNode = J9::TransformUtil::calculateElementAddress(comp(), mhTableNode, arrayIndexNode, TR::Address);
+         TR::Node *mhNode = TR::Node::createWithSymRef(comp()->il.opCodeForIndirectArrayLoad(TR::Address), 1, 1, mhAddressNode, comp()->getSymRefTab()->findOrCreateArrayShadowSymbolRef(TR::Address, mhTableNode));
+         mhNode->copyByteCodeInfo(node);
+         TR::SymbolReference *refinedMHSymRef = comp()->getSymRefTab()->findOrCreateSymRefWithKnownObject(mhNode->getSymbolReference(), mhIndex);
+         mhNode->setSymbolReference(refinedMHSymRef);
```

Comment:
Aliasing copies of symrefs that additionally have a known object index are usually called "improved"